### PR TITLE
Fix: Move docker and docker-compose checking to before method.

### PR DESCRIPTION
**Pull request by saeidrastak**

### CODE REVIEW

File: src/main/java/ir/sahab/dockercomposer/DockerCompose.java [hunk 1]
```diff
@@ -38,15 +37,17 @@ private DockerCompose(List<DockerComposeSpec> specs) {
...
     @Override
     protected void before() throws Throwable {
+        DockerComposeRunner.checkEnvironment();
         // Run the specs one-by-one and collect the services from each spec.
         for (DockerComposeSpec spec : specs) {
+            logger.info("Starting docker-compose spec: {}", spec);
             Map<String, Service> services = spec.runner.start(spec.forceRecreate);
-            // Two services with the same name can exist in these files but docker-compose only run
-            // one of them so we give a light warning here.
+            logger.info("Services information for spec file {} is: {}", spec.file, services);
```
Comment (bozorgzadeh) [inline src/main/java/ir/sahab/dockercomposer/DockerCompose.java]:
```suggestion
            logger.info("A docker-compose with spec {} started. Started services are: {} ", spec, services);
```

File: src/main/java/ir/sahab/dockercomposer/DockerCompose.java [hunk 2]
```diff
@@ -38,15 +37,17 @@ private DockerCompose(List<DockerComposeSpec> specs) {
 
     @Override
     protected void before() throws Throwable {
+        DockerComposeRunner.checkEnvironment();
         // Run the specs one-by-one and collect the services from each spec.
         for (DockerComposeSpec spec : specs) {
+            logger.info("Starting docker-compose spec: {}", spec);
```
Comment (bozorgzadeh) [inline src/main/java/ir/sahab/dockercomposer/DockerCompose.java]:
```suggestion
            logger.info("Starting docker-compose with spec: {}", spec);
```

File: src/main/java/ir/sahab/dockercomposer/DockerCompose.java [hunk 3]
```diff
@@ -56,19 +57,21 @@ protected void before() throws Throwable {
...
+     * behaviour since if the network is used by other services it can not down properly.
      */
     @Override
     protected void after() {
         for (DockerComposeSpec spec : specs) {
+            logger.info("Stopping docker-compose spec: {}", spec);
             if (spec.forceDown) {
                 try {
                     spec.runner.down();
+                    logger.info("Docker-compose spec for file {} stopped!", spec.file.getFileName());
```
Comment (bozorgzadeh) [inline src/main/java/ir/sahab/dockercomposer/DockerCompose.java]:
```suggestion
                    logger.info("A docker-compose with spec {} stopped.", spec);
```

File: src/main/java/ir/sahab/dockercomposer/DockerCompose.java [hunk 4]
```diff
@@ -56,19 +57,21 @@ protected void before() throws Throwable {
...
                 } catch (Exception e) {
                     // We don't care
+                    logger.warn("Unable to stop docker-compose spec: {}", spec);
```
Comment (bozorgzadeh) [inline src/main/java/ir/sahab/dockercomposer/DockerCompose.java]:
```suggestion
                    logger.warn("Unable to stop docker-compose with spec: {}", spec);
```

File: src/main/java/ir/sahab/dockercomposer/DockerCompose.java [hunk 5]
```diff
@@ -108,39 +111,51 @@ public Service getServiceByName(String name) {
...
+        public String toString() {
+            return new ToStringBuilder(this)
+                    .append("file", file)
+                    .append("projectName", projectName)
+                    .append("forceRecreate", forceRecreate)
+                    .append("forceDown", forceDown)
+                    .append("startupCallbacks", startupCallbacks)
+                    .append("runner", runner)
+                    .append("environment", environment)
+                    .toString();
```
Comment (bozorgzadeh) [inline src/main/java/ir/sahab/dockercomposer/DockerCompose.java:134]: Some fields like startupCallbacks, runner and projectName do not provide important information but makes the logs verbose. I think it is better not to include them.

File: src/main/java/ir/sahab/dockercomposer/Service.java [hunk 6]
```diff
@@ -113,4 +105,15 @@ private void copyFrom(Service other) {
...
-}
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this)
+                .append("id", id)
+                .append("name", name)
+                .append("externalIp", externalIp)
+                .append("internalIp", internalIp)
+                .append("portMappings", portMappings)
+                .append("runner", runner)
```
Comment (bozorgzadeh) [inline src/main/java/ir/sahab/dockercomposer/Service.java:116]: Do not provide the runner field please.

File: src/main/java/ir/sahab/dockercomposer/DockerCompose.java (hunk 5 above)
Comment (saeidrastak) [reply src/main/java/ir/sahab/dockercomposer/DockerCompose.java:134]: I think startupCallbacks have good information like WaitFor#PortOpenChecker which is one of the implementations of StartupCallback.

**Pull request merged**